File: app/src/main/java/ru/yourok/utils/SystemParams.java
package ru.yourok.utils;

import android.content.Context;
import android.os.Build;
import android.telephony.TelephonyManager;

import org.json.JSONObject;

import java.io.File;

/**
 * Created by yourok on 19.09.17.
 */

public class SystemParams {
    public static final String DEVICE_ID = "deviceId";
    public static final String DEVICE_TYPE = "deviceType";
    public static final String FIRMWARE = "firmware";
    public static final String MD5_SIGN = "2635881a7ab0593849fe89e685fc56cd";
    public static final String ROOT = "root";
    public static final String SN = "sn";
    public static final String SYSVER = "sysVer";
    public static final String VERSION = "version";
    public static final String IMEI = "imei";

    public static final String U_HOST = "https://u.meizu.com";
    public static final String U_HOST_INTERNATIONAL = "https://u.in.meizu.com";

    private Context context;

    public SystemParams(Context context) {
        this.context = context;
    }

    public JSONObject getSystemParams() throws Exception {
        JSONObject basicParams = new JSONObject();
        basicParams.put(DEVICE_TYPE, getDeviceType());
        basicParams.put(FIRMWARE, getFirmware());
        basicParams.put(ROOT, isRooted());
        basicParams.put(SYSVER, getSysver());
        basicParams.put(VERSION, getSysver());
        basicParams.put(DEVICE_ID, getDeviceId());
        basicParams.put(SN, getSN());
        basicParams.put(IMEI, getDeviceId());
        return basicParams;
    }

    public String getDeviceType() {
        return SystemPropertiesProxy.get(context, "ro.meizu.product.model", "");
    }

    public String getFirmware() {
        return SystemPropertiesProxy.get(context, "ro.build.version.release", "");
    }

    public String getSysver() {
        return SystemPropertiesProxy.get(context, "ro.build.mask.id", "");
    }

    public String isRooted() {
        if (new File("/system/xbin/su").exists())
            return "1";
        return "0";
    }

    public String getDeviceId() {//imei
        TelephonyManager telephonyManager;
        telephonyManager = (TelephonyManager) context.getSystemService(Context.TELEPHONY_SERVICE);
        return telephonyManager.getDeviceId();
    }

    public String getSN() {
        return Build.SERIAL;
    }

    static public String getHost(boolean international) {
        if (international)
            return U_HOST_INTERNATIONAL;
        return U_HOST;
    }

    static public String getSign(String params) {
        return MD5(params + MD5_SIGN);
    }


    static public String MD5(String md5) {
        try {
            java.security.MessageDigest md = java.security.MessageDigest.getInstance("MD5");
            byte[] array = md.digest(md5.getBytes());
            StringBuffer sb = new StringBuffer();
            for (int i = 0; i < array.length; ++i) {
                sb.append(Integer.toHexString((array[i] & 0xFF) | 0x100).substring(1, 3));
            }
            return sb.toString();
        } catch (java.security.NoSuchAlgorithmException e) {
        }
        return null;
    }
}

File: app/src/main/java/ru/yourok/utils/SystemParams.kt
package ru.yourok.utils

import android.content.Context
import android.os.Build
import android.telephony.TelephonyManager
import org.json.JSONObject
import java.io.File
import java.security.MessageDigest

/**
 * Created by yourok on 19.09.17.
 */

class SystemParams(private val context: Context) {

    val systemParams: JSONObject
        @Throws(Exception::class)
        get() {
            val basicParams = JSONObject()
            basicParams.put(DEVICE_TYPE, deviceType)
            basicParams.put(FIRMWARE, firmware)
            basicParams.put(ROOT, isRooted)
            basicParams.put(SYSVER, sysver)
            basicParams.put(VERSION, sysver)
            basicParams.put(DEVICE_ID, deviceId)
            basicParams.put(SN, sn)
            basicParams.put(IMEI, deviceId)
            return basicParams
        }

    val deviceType: String
        get() = SystemPropertiesProxy.get(context, "ro.meizu.product.model", "")

    val firmware: String
        get() = SystemPropertiesProxy.get(context, "ro.build.version.release", "")

    val sysver: String
        get() = SystemPropertiesProxy.get(context, "ro.build.mask.id", "")

    val isRooted: String
        get() = if (File("/system/xbin/su").exists()) "1" else "0"

    //imei
    val deviceId: String
        get() {
            try {
                val telephonyManager: TelephonyManager
                telephonyManager = context.getSystemService(Context.TELEPHONY_SERVICE) as TelephonyManager
                return telephonyManager.deviceId
            } catch (e: Exception) {
                e.printStackTrace()
                return ""
            }

        }

    val sn: String
        get() = Build.SERIAL

    companion object {
        val DEVICE_ID = "deviceId"
        val DEVICE_TYPE = "deviceType"
        val FIRMWARE = "firmware"
        val MD5_SIGN = "2635881a7ab0593849fe89e685fc56cd"
        val ROOT = "root"
        val SN = "sn"
        val SYSVER = "sysVer"
        val VERSION = "version"
        val IMEI = "imei"

        val U_HOST = "https://u.meizu.com"
        val U_HOST_INTERNATIONAL = "https://u.in.meizu.com"

        fun getHost(international: Boolean): String {
            return if (international) U_HOST_INTERNATIONAL else U_HOST
        }

        fun getSign(params: String): String? {
            return MD5(params + MD5_SIGN)
        }


        fun MD5(md5: String): String? {
            try {
                val sb = StringBuffer()
                val md = MessageDigest.getInstance("MD5")
                val digest = md.digest(md5.toByteArray())
                for (byte in digest) sb.append("%02x".format(byte))
                return sb.toString()
            } catch (e: java.security.NoSuchAlgorithmException) {
            }

            return null
        }
    }
}